pkgs/native_assets_cli/lib/src/api/c_code_asset.dart

```diff
@@ -0,0 +1,100 @@
+// Copyright (c) 2024, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+part of 'asset.dart';
+
+/// A code [Asset] which respects the C application binary interface (ABI).
+///
+/// Typical other languages which produce code assets that respect the C ABI
+/// include C++ and Rust.
+///
+/// There are several types of C code assets:
+/// * Assets which designate symbols present in the target system
+///   ([SystemDylib]), process ([LookupInProcess]), or executable
+///   ([LookupInExecutable]). These assets are identified by their [id], and do
+///   not have a [file].
+/// * Dynamic libraries bundled into the application ([BundledDylib]). These
+///   assets must provide a [file] to be bundled.
+///
+/// An application is compiled to run on a specific target [os] and
+/// [architecture]. Different targets require different assets, so the package
+/// developer must specify which asset to bundle for which target.
+///
+/// An asset has different ways of being accessible in the final application. It
+/// is either brought in "manually" by having the package developer specify a
+/// [file] path of the asset on the current system, it can be part of the Dart
+/// or Flutter SDK ([LookupInProcess]), or it can be already present in the
+/// target system ([SystemDylib]). If the asset is bundled "manually", the Dart
+/// or Flutter SDK will take care of copying the asset [file] from its specified
+/// location on the current system into the application bundle.
+///
+/// Assets are also called "native assets" to differentiate them from the Dart
+/// code also bundled with an application.
+abstract final class CCodeAsset implements Asset {
+  /// The operating system this asset can run on.
+  OS get os;
+
+  /// The architecture this asset can run on.
+  ///
+  /// Not available during a [BuildConfig.dryRun].
+  Architecture? get architecture;
+
+  /// The link mode for this native code.
+  ///
+  /// Either dynamic loading or static linking.
+  LinkMode get linkMode;
+
+  /// The dynamic loading method when the [linkMode] is [LinkMode.dynamic].
+  DynamicLoading get dynamicLoading;
```

Comment:
What is the value of this when the `linkMode` is not `dynamic`?

Reply:
Hm, we need to decide on a nice API with such alternatives.

It's kind of the same thing as with some of the fields not being present in `dryRun` in `BuildConfig` but guaranteed to be present in non-`dryRun`.
This field must be present if `linkMode == LinkMode.dynamic` but not if `linkMode == LinkMode.static`.

I considered trying to merge `LinkMode` and `DynamicLoading` into a single type and single field, but that kind of conflates two decisions.

@lrhn Any bright ideas? Do we have prior art of when fields are not accessible? Should it be non-nullable but throwing `StateError`s if accessed?

Reply:
Sounds like a property of the `LinkMode`, so a field of `LinkMode.dynamic` (or an extension of that), to me.